src/v/cluster/partition.cc

```diff
@@ -414,6 +416,28 @@ ss::future<> partition::remove_persistent_state() {
     if (_id_allocator_stm) {
         co_await _id_allocator_stm->remove_persistent_state();
     }
+
+    // Backward compatibility: even if remote.delete is true, only do
+    // deletion if the partition is in full tiered storage mode (this
+    // excludes read replica clusters from deleting data in S3)
+    bool tiered_storage
+      = get_ntp_config().is_tiered_storage()
+        || (config::shard_local_cfg().cloud_storage_enable_remote_write() &&
+            config::shard_local_cfg().cloud_storage_enable_remote_read());
+
+    if (_cloud_storage_partition && tiered_storage && _remote_delete_enabled) {
+        vlog(
+          clusterlog.debug,
+          "Erasing S3 objects for partition {} ({} {} {})",
+          ntp(),
+          get_ntp_config(),
+          get_ntp_config().is_archival_enabled(),
+          get_ntp_config().is_read_replica_mode_enabled());
+        co_await _cloud_storage_partition->erase();
+    } else {
+        vlog(
+          clusterlog.info, "Leaving S3 objects behind for partition {}", ntp());
+    }
```

Comment:
I am not sure if this code should be part of `remove_persistent_state` this method is called when a partition is moved between nodes (it previously removed only local state) Now it will remove data from object store